transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java

```diff
@@ -0,0 +1,170 @@
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.hydrator.plugin;
+
+import co.cask.cdap.api.annotation.Description;
+import co.cask.cdap.api.annotation.Name;
+import co.cask.cdap.api.annotation.Plugin;
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.api.plugin.PluginConfig;
+import co.cask.cdap.etl.api.Emitter;
+import co.cask.cdap.etl.api.PipelineConfigurer;
+import co.cask.cdap.etl.api.Transform;
+import co.cask.cdap.etl.api.TransformContext;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.google.common.collect.Lists;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Transforms records by normalizing the data.
+ * Convert wide rows and reducing data to it canonicalize form
+ */
+@Plugin(type = Transform.PLUGIN_TYPE)
+@Name("Normalize")
+@Description("Convert wide rows and reducing data to it canonicalize form")
+public class Normalize extends Transform<StructuredRecord, StructuredRecord> {
+  public static final String NAME_KEY_SUFFIX = "_name";
+  public static final String VALUE_KEY_SUFFIX = "_value";
+
+  private final NormalizeConfig config;
+
+  private Schema outputSchema;
+  private Map<String, String> mappingFieldMap;
+  private Map<String, String> normalizeFieldMap;
+  private List<String> normalizeFieldList;
+
+  public Normalize(NormalizeConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) throws IllegalArgumentException {
+    super.configurePipeline(pipelineConfigurer);
+    //validate fields with input schema
+    Schema inputSchema = pipelineConfigurer.getStageConfigurer().getInputSchema();
+    //validate mapping fields
+    String[] fieldMappingArray = config.fieldMapping.split(",");
+    for (String fieldMapping : fieldMappingArray) {
+      String[] mappings = fieldMapping.split(":");
+      Preconditions.checkArgument(inputSchema.getField(mappings[0]) != null, "Mapping field '" + mappings[0]
+        + "' not present in input schema.");
+    }
+    //validate normalizing fields
+    String[] fieldNormalizingArray = config.fieldNormalizing.split(",");
+    for (String fieldNormalizing : fieldNormalizingArray) {
+      String[] fields = fieldNormalizing.split(":");
+      Preconditions.checkArgument(inputSchema.getField(fields[0]) != null, "Normalizing field '" + fields[0]
```

Comment:
Tested on hydratory UI, widget does not allow null or empty value, hence cannot be null.